exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,305 @@
+                                            async_mqtt::allocate_buffer(mqtt_password_),
+                                            { async_mqtt::property::session_expiry_interval{ UINT_MAX } } };
+
+        co_await mqtt_client_->send(connect_packet, asio::use_awaitable);
+
+        logger_.trace("Waiting for MQTT connection acknowledgement");
+        co_await mqtt_client_->recv(async_mqtt::filter::match, { async_mqtt::control_packet_type::connack },
+                                    asio::use_awaitable);
+
+        connect_active_ = false;
+        co_return;
+      } catch (std::exception& e) {
+        logger_.error("Error while connecting to MQTT broker: {}", e.what());
+      } catch (...) {
+        logger_.error("Unknown error while connecting to MQTT broker");
+      }
+
+      // If the connection failed, then clear the socket and try again after 1 second.
+      logger_.trace("Waiting for 1 second and then trying to connect again");
+      co_await asio::co_spawn(mqtt_client_->strand(), mqtt_client_->close(asio::use_awaitable), asio::use_awaitable);
+      co_await asio::steady_timer{ ctx_, std::chrono::seconds{ 1 } }.async_wait(asio::use_awaitable);
+    }
+  }
+
+  // Function which has oversight over the signals
+  auto load_signals() -> void {
+    // Stop reading the current signals by canceling the slots
+    for (auto& slot : slots_) {
+      std::visit([](auto& ptr) { ptr->cancel(); }, slot);
+    }
+
+    active_signals_.clear();
+    get_signals();
+    clean_signals();
+
+    for (tfc::ipc_ruler::signal& signal : active_signals_) {
+      handle_signal(signal);
+    }
+  }
+
+  auto get_signals() -> void {
+    ipc_client_.signals([&](const std::vector<tfc::ipc_ruler::signal>& signals) {
+      for (const tfc::ipc_ruler::signal& signal : signals) {
+        active_signals_.push_back(signal);
+      }
+    });
+
+    ctx_.run_for(std::chrono::milliseconds(20));
+  }
+
+  // This function removes signals that are banned
+  auto clean_signals() -> void {
+    active_signals_.erase(std::remove_if(active_signals_.begin(), active_signals_.end(),
+                                         [&](const tfc::ipc_ruler::signal& signal) {
+                                           return std::ranges::any_of(banned_signals_.begin(), banned_signals_.end(),
+                                                                      [&](const std::string& banned_string) {
+                                                                        return signal.name.find(banned_string) !=
+                                                                               std::string::npos;
+                                                                      });
+                                         }),
+                          active_signals_.end());
+  }
+
+  // This function checks the type of the signal in order to determine how to read from the slot
+  auto handle_signal(tfc::ipc_ruler::signal& signal) -> void {
```

Comment:
we have similar code in impl.hpp of ipc, could we reuse that code or change it somehow to being more reusable

Reply:
I cannot find similar code in impl.hpp, should this function be moved to be inside of impl.hpp?

Reply:
a similar behavior is on the bottom of the impl.hpp

Reply:
this code has changed significantly and types are declared automatically, so this issue has been resolved.